src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,107 @@
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    size_t k{_secret_hash(user.name + user.pass)};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                co_await item.client->stop();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        // Otherwise the user has a client
+        vlog(plog.debug, "Reuse client for user {}", k);
+
+        // Grab the node located at that position
+        auto node = inner_hash.extract(it_hash);
+
+        // Update the timestamp and put it to the front
+        node.value().last_used = timestamped_user::clock::now();
+        client = node.value().client;
+        inner_list.insert(inner_list.begin(), std::move(node));
+    }
+
+    co_return co_await ss::make_ready_future<client_ptr>(client);
+}
+
+void kafka_client_cache::clean_stale_clients() {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto now = timestamped_user::clock::now();
+    inner_list.remove_if([this, now](const timestamped_user& item) {
+        bool remove = (now >= item.last_used + _keep_alive());
+        if (remove) {
+            vlog(plog.debug, "Removed {} from cache", item.key);
+        }
+        return remove;
+    });
+}
+
+ss::future<> kafka_client_cache::stop() {
+    auto& inner_list = _cache.get<underlying_list>();
+    for (auto it = inner_list.cbegin(); it != inner_list.cend(); it++) {
+        co_await it->client->stop();
+    }
```

Comment:
Does this need to guard against the cache being modified whilst this loop is asynchronously iterating?

There are a couple of techniques I can think of:
* Swap with an empty one on the stack.
* Use while(!empty) and remove an item each iteration

Reply:
This stop will fire when the cache is stopped at app termination.
So using the while loop should be sufficient because it will account for any lingering cache insertions from other requests